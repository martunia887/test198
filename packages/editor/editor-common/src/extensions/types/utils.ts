import { ExtensionNode } from './extension-manifest';

export type MenuItem = {
  key: string;
  title: string;
  description: string;
  icon: () => Promise<any>;
  node: ExtensionNode;
};

export type MenuItemMap = { [key: string]: MenuItem };
