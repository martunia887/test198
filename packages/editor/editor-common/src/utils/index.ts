export * from './color';
export * from './macro';
export * from './validator';
export { default as browser } from './browser';
export {
  default as ErrorReporter,
  ErrorReportingHandler,
} from './error-reporter';
export * from './date';
export * from './imageLoader';
export * from './calc-breakout-width';
export { default as ADFTraversor } from './traversor';
export * from './analytics';
export { measureRender } from './performance/measure-render';
export { startMeasure, stopMeasure, clearMeasure } from './performance/measure';
export { getResponseEndTime } from './performance/navigation';
