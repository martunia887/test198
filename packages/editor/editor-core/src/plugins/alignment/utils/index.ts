import { findParentNodeOfType } from 'prosemirror-utils';
import { CellSelection } from 'prosemirror-tables';
import { AlignmentState } from '../pm-plugins/main';

export const getActiveAlignment = (state): AlignmentState | undefined => {
  if (state.selection instanceof CellSelection) {
    let marks: string[] = [];
    state.selection.forEachCell(cell => {
      const mark = cell.firstChild!.marks.filter(
        mark => mark.type === state.schema.marks.alignment,
      )[0];
      if (mark) {
        marks.push(mark.attrs.align);
      } else {
        marks.push('start');
      }
    });

    let seen;
    for (let i = 0, len = marks.length; i < len; i++) {
      const mark = marks[i];

      if (typeof seen === 'undefined') {
        seen = mark;
      } else if (seen !== mark) {
        seen = null;
        break;
      }
    }
    return seen ? seen : 'start';
  }

  const node = findParentNodeOfType([
    state.schema.nodes.paragraph,
    state.schema.nodes.heading,
  ])(state.selection);
  const getMark =
    node &&
    node.node.marks.filter(
      mark => mark.type === state.schema.marks.alignment,
    )[0];

  return (getMark && getMark.attrs.align) || 'start';
};
