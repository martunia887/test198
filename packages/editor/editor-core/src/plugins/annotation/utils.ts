import { Node } from 'prosemirror-model';
import { EditorState, TextSelection, Transaction } from 'prosemirror-state';

export function canAnnotationBeCreatedInRange(from: number, to: number) {
  return (state: EditorState) => {
    const {
      schema: {
        marks: { annotation: annotationMark },
      },
    } = state;

    if (!state.doc.rangeHasMark(from, to, annotationMark)) {
      const $from = state.doc.resolve(from);
      const $to = state.doc.resolve(to);

      if (
        $from.parent === $to.parent &&
        $from.parent.isTextblock &&
        $from.parent.type.allowsMarkType(annotationMark)
      ) {
        let allowed = true;
        state.doc.nodesBetween(from, to, node => {
          allowed =
            allowed && !node.marks.some(m => m.type.excludes(annotationMark));
          return allowed;
        });
        return allowed;
      }
    }

    return false;
  };
}

export function isSelectionInsideAnnotation(
  stateOrTr: EditorState | Transaction,
) {
  return !!stateOrTr.doc.type.schema.marks.annotation.isInSet(
    stateOrTr.selection.$from.marks(),
  );
}

export function isSelectionAroundAnnotation(
  stateOrTr: EditorState | Transaction,
) {
  const {
    selection: { $from, $to },
  } = stateOrTr;
  const node = $from.nodeAfter;

  return (
    !!node &&
    $from.textOffset === 0 &&
    $to.pos - $from.pos === node.nodeSize &&
    !!stateOrTr.doc.type.schema.marks.annotation.isInSet(node.marks)
  );
}

export function getActiveAnnotationMark(
  state: EditorState | Transaction,
): { node: Node; pos: number } | undefined {
  const {
    selection: { $from },
  } = state;

  if (
    isSelectionInsideAnnotation(state) ||
    isSelectionAroundAnnotation(state)
  ) {
    const pos = $from.pos - $from.textOffset;
    const node = state.doc.nodeAt(pos) as Node;
    return node && node.isText ? { node, pos } : undefined;
  }

  return undefined;
}

export function getActiveText(state: EditorState): string | undefined {
  const { selection, tr } = state;

  if (selection instanceof TextSelection) {
    const { $from, from, to } = selection;
    const node = tr.doc.nodeAt(from);

    if (node && node.isText) {
      return node.textContent.substr($from.textOffset, to - from);
    }
  }

  return;
}
