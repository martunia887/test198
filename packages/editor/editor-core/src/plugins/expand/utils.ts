import { EditorState } from 'prosemirror-state';
import {
  findSelectedNodeOfType,
  findParentNodeOfType,
} from 'prosemirror-utils';

export const findExpand = (state: EditorState) => {
  const { expand, nestedExpand } = state.schema.nodes;
  return (
    findSelectedNodeOfType([expand, nestedExpand])(state.selection) ||
    findParentNodeOfType([expand, nestedExpand])(state.selection)
  );
};
