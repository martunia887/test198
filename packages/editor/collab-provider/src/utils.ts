import { AllSelection, NodeSelection, Selection } from 'prosemirror-state';

import { SendableSelection } from './types';

function isAllSelection(selection: Selection) {
  return selection instanceof AllSelection;
}

function isNodeSelection(selection: Selection) {
  return selection instanceof NodeSelection;
}

export const getSendableSelection = (
  selection: Selection,
): SendableSelection => {
  /**
   * <kbd>CMD + A</kbd> triggers a AllSelection
   * <kbd>escape</kbd> triggers a NodeSelection
   */
  return {
    type: 'textSelection',
    anchor: selection.anchor,
    head:
      isAllSelection(selection) || isNodeSelection(selection)
        ? selection.head - 1
        : selection.head,
  };
};

const participants = [
  'Awilda Mcentire',
  'Katie Iddings',
  'Serafina Mcdowell',
  'Eun Knopp',
  'Walker Reighard',
  'Keely Muth',
  'Allie April',
  'Rosamond Paton',
  'Lashawnda Peppard',
  'Hubert Feingold',
  'Idalia Mcquaig',
  'Wyatt Byer',
  'Genevieve Reeder',
  'Dona Elton',
  'Jerilyn Holte',
  'Salina Erhardt',
  'Velda Latour',
  'Regina Richer',
  'Natisha Tiger',
  'Glynda Stuckey',
  'Wilson Woodberry',
  'Elvie Furtado',
  'Lasandra Palmatier',
  'Louise Mijares',
  'Herma Vanleer',
  'Titus Plante',
  'James Fiscus',
  'Ina Ramero',
  'Jeanene Bernabe',
  'Lida Lepine',
  'Rosalina Palka',
  'Dayle Grossi',
  'Zack Leister',
  'Isobel Fay',
  'Elouise Kluge',
  'Twyla Machin',
  'Diane Herrington',
  'Fumiko Elders',
  'Kennith Howe',
  'Doloris Mccollom',
  'Daryl Wherry',
  'Sade Albury',
  'Roy Cork',
  'Hunter Hultgren',
  'Mi Miley',
  'Isabelle Halperin',
  'Kayla Starke',
  'Allyson Woomer',
  'Sirena Politte',
  'Patricia Scovil',
];

function hashCode(s: string) {
  let h = 0;
  for (let i = 0; i < s.length; i++)
    h = (Math.imul(31, h) + s.charCodeAt(i)) | 0;

  return Math.abs(h);
}

export const getParticipant = (userId: string) => {
  // const id = +userId.split('/')[1];
  // console.log({userId});
  const id = hashCode(userId);
  // eslint-disable-next-line no-bitwise
  const name = participants[(id | 0) % participants.length];
  return {
    userId,
    name,
    avatar: `https://api.adorable.io/avatars/80/${name.replace(/\s/g, '')}.png`,
    email: `${name.replace(/\s/g, '').toLocaleLowerCase()}@atlassian.com`,
  };
};

export const createLogger = (prefix: string, color: string = 'blue') => (
  msg: string,
  data: any = null,
) => {
  // eslint-disable-next-line no-console
  console.log(`%c${prefix}: ${msg}`, `color: ${color}; font-weight: bold`);
  if (data) {
    // eslint-disable-next-line no-console
    console.log(data);
  }
};
