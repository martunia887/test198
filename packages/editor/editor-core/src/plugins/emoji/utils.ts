import { EditorState, StateField, PluginKey } from 'prosemirror-state';
import { Command } from '../../types';
import { Dispatch } from '../../event-dispatcher';

export type Reducer<S, A> = (state: S, action: A) => S;

export function pluginStateFactory<S, A, IS extends S>(
  pluginKey: PluginKey,
  initialState: IS,
  reducer: Reducer<S, A>,
): {
  createState: (dispatch: Dispatch) => StateField<S>;
  createCommand: (action: A) => Command;
  getPluginState: (editorState: EditorState) => S;
} {
  return {
    createState: dispatch => ({
      init: () => initialState,
      apply(tr, pluginState) {
        const meta = tr.getMeta(pluginKey);
        if (meta) {
          const newState = reducer(pluginState, meta);
          dispatch(pluginKey, newState);
          return newState;
        }
        return pluginState;
      },
    }),
    createCommand: (action): Command => (state, dispatch) => {
      dispatch(state.tr.setMeta(pluginKey, action));
      return true;
    },
    getPluginState: editorState => pluginKey.getState(editorState),
  };
}
