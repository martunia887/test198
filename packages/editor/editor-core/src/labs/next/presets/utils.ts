import { EditorPresetProps, PluginsPreset } from './types';
import { EditorPlugin } from '../../../types';
import { compose } from '../../../utils';

export const removeExcludes = (excludes: EditorPresetProps['excludes']) => (
  plugins: EditorPlugin[],
) => {
  if (excludes) {
    return plugins.filter(plugin => excludes.indexOf(plugin.name) === -1);
  }
  return plugins;
};

export type ExperimentalPluginMap = Map<string, EditorPlugin>;
export const enableExperimental = (
  experimental: EditorPresetProps['experimental'],
  experimentalPluginMap: ExperimentalPluginMap,
) => (plugins: EditorPlugin[]) => {
  if (experimental && experimental.length) {
    experimental.map(pluginName => {
      const plugin = experimentalPluginMap.get(pluginName);
      if (plugin) {
        plugins.push(plugin);
      }
    });
  }
  return plugins;
};

export const processPluginPreset = (preset: PluginsPreset) => {
  const cache = new Map();
  preset.forEach(pluginEntry => {
    if (Array.isArray(pluginEntry)) {
      const [fn, options] = pluginEntry;
      cache.set(fn, options);
    } else {
      /**
       * Prevent usage of same plugin twice without override.
       * [
       *  plugin1,
       *  [plugin1, { option1: value }],
       *  plugin1, // This will throw
       * ]
       */
      if (cache.has(pluginEntry) && cache.get(pluginEntry) === undefined) {
        throw new Error(`${pluginEntry} is already included!`);
      }
      cache.set(pluginEntry, undefined);
    }
  });

  let plugins: Array<EditorPlugin> = [];
  cache.forEach((options, fn) => {
    plugins.push(fn(options));
  });

  return plugins;
};

export const getPluginsFromPreset = (
  preset: PluginsPreset,
  excludes?: EditorPresetProps['excludes'],
  experimental?: EditorPresetProps['experimental'],
): Array<EditorPlugin> => {
  const experimentalMap: ExperimentalPluginMap = new Map();

  return compose(
    enableExperimental(experimental, experimentalMap),
    removeExcludes(excludes),
  )(processPluginPreset(preset));
};
