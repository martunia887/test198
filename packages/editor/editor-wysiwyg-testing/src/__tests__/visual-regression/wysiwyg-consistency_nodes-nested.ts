import { traverse } from '@atlaskit/adf-utils/traverse';
import { createDocumentADF } from './utils/_adf-utils';
import { loadKitchenSinkWithAdf } from './utils/_example-utils';
import { snapshotAndCompare } from './utils/_comparison-utils';
import { FragmentNodeLookup, contentNodes, containerNodes } from './_data';

describe('WYSIWYG Snapshot Test: Nested nodes look consistent in editor & renderer', () => {
  let page: any;

  beforeAll(async () => {
    // @ts-ignore
    page = global.page;
    await page.setViewport({ width: 2000, height: 1000 });
  });

  describe.each(containerNodes)(
    '%p',
    ({ node: containerNode }: FragmentNodeLookup) => {
      const containerAdf = createDocumentADF(containerNode, true);
      // Nested nodes
      it.each(contentNodes)(
        `%p inside ${containerNode}`,
        async ({ node: contentNode, waitFor }: FragmentNodeLookup) => {
          const contentAdf = createDocumentADF(contentNode).content;
          const adf = traverse(containerAdf, {
            any: (node: any) => {
              if (node.content && node.content.length === 0) {
                // Insert nested content into container
                node.content = contentAdf;
              }
              return node;
            },
          });
          await loadKitchenSinkWithAdf(page, adf);
          await snapshotAndCompare(
            page,
            `${contentNode} inside ${containerNode}`,
            waitFor,
          );
        },
      );
    },
  );
});
