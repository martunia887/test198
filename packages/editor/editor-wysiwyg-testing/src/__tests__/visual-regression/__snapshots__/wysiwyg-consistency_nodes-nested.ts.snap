// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WYSIWYG Comparison: actions inside columns 1`] = `
Object {
  "consistency": "92.05%",
  "divergence": 0.0794590856144438,
}
`;

exports[`WYSIWYG Comparison: actions inside panel 1`] = `
Object {
  "consistency": "88.85%",
  "divergence": 0.11146304675716441,
}
`;

exports[`WYSIWYG Comparison: actions inside table 1`] = `
Object {
  "consistency": "95.1%",
  "divergence": 0.04896861429135413,
}
`;

exports[`WYSIWYG Comparison: blockquote inside columns 1`] = `
Object {
  "consistency": "84.67%",
  "divergence": 0.1533067464273845,
}
`;

exports[`WYSIWYG Comparison: blockquote inside panel 1`] = `
Object {
  "consistency": "88.85%",
  "divergence": 0.11146304675716441,
}
`;

exports[`WYSIWYG Comparison: blockquote inside table 1`] = `
Object {
  "consistency": "79.3%",
  "divergence": 0.20697448096885812,
}
`;

exports[`WYSIWYG Comparison: bullet list inside columns 1`] = `
Object {
  "consistency": "97.23%",
  "divergence": 0.027658045977011495,
}
`;

exports[`WYSIWYG Comparison: bullet list inside panel 1`] = `
Object {
  "consistency": "99.99%",
  "divergence": 0.00014450867052023122,
}
`;

exports[`WYSIWYG Comparison: bullet list inside table 1`] = `
Object {
  "consistency": "98.36%",
  "divergence": 0.016435415974299324,
}
`;

exports[`WYSIWYG Comparison: codeblock inside columns 1`] = `
Object {
  "consistency": "96.11%",
  "divergence": 0.038887347391786904,
}
`;

exports[`WYSIWYG Comparison: codeblock inside panel 1`] = `
Object {
  "consistency": "99.99%",
  "divergence": 0.00014450867052023122,
}
`;

exports[`WYSIWYG Comparison: codeblock inside table 1`] = `
Object {
  "consistency": "97.17%",
  "divergence": 0.028300070028011205,
}
`;

exports[`WYSIWYG Comparison: date inside columns 1`] = `
Object {
  "consistency": "94.1%",
  "divergence": 0.05897435897435897,
}
`;

exports[`WYSIWYG Comparison: date inside panel 1`] = `
Object {
  "consistency": "99%",
  "divergence": 0.01001589825119237,
}
`;

exports[`WYSIWYG Comparison: date inside table 1`] = `
Object {
  "consistency": "95.27%",
  "divergence": 0.04728213507625272,
}
`;

exports[`WYSIWYG Comparison: decisions inside columns 1`] = `
Object {
  "consistency": "91.57%",
  "divergence": 0.08433677926616191,
}
`;

exports[`WYSIWYG Comparison: decisions inside panel 1`] = `
Object {
  "consistency": "99%",
  "divergence": 0.01001589825119237,
}
`;

exports[`WYSIWYG Comparison: decisions inside table 1`] = `
Object {
  "consistency": "94.93%",
  "divergence": 0.05066127121989736,
}
`;

exports[`WYSIWYG Comparison: divider inside columns 1`] = `
Object {
  "consistency": "100%",
  "divergence": 0,
}
`;

exports[`WYSIWYG Comparison: divider inside panel 1`] = `
Object {
  "consistency": "99%",
  "divergence": 0.01001589825119237,
}
`;

exports[`WYSIWYG Comparison: divider inside table 1`] = `
Object {
  "consistency": "97.7%",
  "divergence": 0.02303921568627451,
}
`;

exports[`WYSIWYG Comparison: emoji inside columns 1`] = `
Object {
  "consistency": "91.27%",
  "divergence": 0.08729260935143288,
}
`;

exports[`WYSIWYG Comparison: emoji inside panel 1`] = `
Object {
  "consistency": "98.9%",
  "divergence": 0.010969793322734499,
}
`;

exports[`WYSIWYG Comparison: emoji inside table 1`] = `
Object {
  "consistency": "94.24%",
  "divergence": 0.05759259259259259,
}
`;

exports[`WYSIWYG Comparison: heading inside columns 1`] = `
Object {
  "consistency": "91.31%",
  "divergence": 0.0868976846057572,
}
`;

exports[`WYSIWYG Comparison: heading inside panel 1`] = `
Object {
  "consistency": "99.04%",
  "divergence": 0.009599224305106659,
}
`;

exports[`WYSIWYG Comparison: heading inside table 1`] = `
Object {
  "consistency": "95.14%",
  "divergence": 0.04861150770988007,
}
`;

exports[`WYSIWYG Comparison: mention inside columns 1`] = `
Object {
  "consistency": "94.95%",
  "divergence": 0.050490196078431374,
}
`;

exports[`WYSIWYG Comparison: mention inside panel 1`] = `
Object {
  "consistency": "99.42%",
  "divergence": 0.005802861685214627,
}
`;

exports[`WYSIWYG Comparison: mention inside table 1`] = `
Object {
  "consistency": "95.7%",
  "divergence": 0.042984749455337694,
}
`;

exports[`WYSIWYG Comparison: ordered list inside columns 1`] = `
Object {
  "consistency": "96.92%",
  "divergence": 0.030802062204192023,
}
`;

exports[`WYSIWYG Comparison: ordered list inside panel 1`] = `
Object {
  "consistency": "99.99%",
  "divergence": 0.00014450867052023122,
}
`;

exports[`WYSIWYG Comparison: ordered list inside table 1`] = `
Object {
  "consistency": "98.2%",
  "divergence": 0.01796831727041099,
}
`;

exports[`WYSIWYG Comparison: paragraph inside columns 1`] = `
Object {
  "consistency": "83.65%",
  "divergence": 0.16347402597402597,
}
`;

exports[`WYSIWYG Comparison: paragraph inside panel 1`] = `
Object {
  "consistency": "93.61%",
  "divergence": 0.06387254901960784,
}
`;

exports[`WYSIWYG Comparison: paragraph inside table 1`] = `
Object {
  "consistency": "80.01%",
  "divergence": 0.19986663783160322,
}
`;

exports[`WYSIWYG Comparison: status inside columns 1`] = `
Object {
  "consistency": "88.85%",
  "divergence": 0.11146304675716441,
}
`;

exports[`WYSIWYG Comparison: status inside panel 1`] = `
Object {
  "consistency": "99.49%",
  "divergence": 0.005127186009538951,
}
`;

exports[`WYSIWYG Comparison: status inside table 1`] = `
Object {
  "consistency": "94.76%",
  "divergence": 0.05239651416122004,
}
`;
